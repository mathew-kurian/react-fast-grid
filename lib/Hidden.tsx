import * as React from "react";
import { Breakpoint, keys } from "./createBreakPoints";
import { breakpoints } from "./gridHelpers";
import debounce from "./debounce";

interface HiddenState {
  visible: boolean;
}

const reflect = (willHide: boolean) => willHide;

export interface HiddenProps {
  xsUp?: boolean;
  xsDown?: boolean;
  smUp?: boolean;
  smDown?: boolean;
  mdUp?: boolean;
  mdDown?: boolean;
  lgUp?: boolean;
  lgDown?: boolean;
  xlUp?: boolean;
  xlDown?: boolean;
  predicate?: (willHide: boolean) => boolean;
  only?: Breakpoint | Breakpoint[];
  debounce?: number;
  children?: any;
}

const upFrom = (breakpoint: Breakpoint): Breakpoint[] => {
  return keys.slice(keys.indexOf(breakpoint) + 1);
};

const downFrom = (breakpoint: Breakpoint): Breakpoint[] => {
  return keys.slice(0, keys.indexOf(breakpoint));
};

type VisibilityMap = {
  [key in keyof Required<
    Omit<HiddenProps, "children" | "only" | "debounce" | "predicate">
  >]: Breakpoint[];
};

export const VisibilityMap: VisibilityMap = {
  xsUp: downFrom("xs"),
  xsDown: upFrom("xs"),
  smUp: downFrom("sm"),
  smDown: upFrom("sm"),
  mdUp: downFrom("md"),
  mdDown: upFrom("md"),
  lgUp: downFrom("lg"),
  lgDown: upFrom("lg"),
  xlUp: downFrom("xl"),
  xlDown: upFrom("xl"),
};

export default class Hidden extends React.Component<HiddenProps, HiddenState> {
  private static keysDsc = [...keys].reverse();
  private static init_: boolean = false;
  private static listeners_: (() => void)[] = [];

  private static init() {
    if (Hidden.init_) {
      return;
    }

    Hidden.init_ = true;

    if (typeof window.matchMedia === "function") {
      for (const [i, key] of keys.entries()) {
        const start = key;
        const end = keys[i + 1];

        const media = window.matchMedia(
          breakpoints.between(start, end).replace("@media", "").trim()
        );

        media.addListener(() => {
          if (process.env.NODE_ENV !== "production") {
            console.log("Fired listener for media items");
          }

          for (const listener of Hidden.listeners_) {
            listener();
          }
        });
      }
    } else {
      window.addEventListener(
        "resize",
        () => {
          for (const listener of Hidden.listeners_) {
            listener();
          }
        },
        false
      );
    }
  }

  private _onResizeDebounced: () => void;

  constructor(props: HiddenProps) {
    super(props);

    Hidden.init();

    const width = window.innerWidth;
    const breakPoint = this._getBreakPoint(width);

    this.state = {
      visible: this._isVisible(breakPoint),
    };

    const { debounce = 5 } = props;

    this._onResizeDebounced = this._getSafeDebounceResize(
      this._onResize,
      debounce
    );

    if (process.env.NODE_ENV !== "production") {
      const { predicate, ...otherProps } = props;

      if (Object.keys(otherProps).length > 1) {
        console.warn(
          "[Hidden] is recommended to have only 1 prop except for `predicate` which can be in conjunction with another prop"
        );
      }
    }
  }

  private _getSafeDebounceResize<T extends Function>(
    callback: T,
    timeout: number | undefined
  ): T {
    return typeof timeout === "number" && timeout > 0
      ? debounce(callback, timeout)
      : callback;
  }

  // support React 15+
  shouldComponentUpdate(nextProps: HiddenProps) {
    const { debounce = 5 } = nextProps;

    if (this.props.debounce !== debounce) {
      this._onResizeDebounced = this._getSafeDebounceResize(
        this._onResize,
        debounce
      );
    }

    return true;
  }

  componentDidMount() {
    Hidden.listeners_.push(this._onResizeInvoker);
  }

  componentWillUnmount() {
    Hidden.listeners_ = Hidden.listeners_.filter(
      (listener) => listener !== this._onResizeInvoker
    );
  }

  private _getBreakPoint = (width: number): Breakpoint => {
    const breakPointValues = breakpoints.values;

    for (const key of Hidden.keysDsc) {
      if (width >= breakPointValues[key]) {
        return key;
      }
    }

    return "xl";
  };

  private _getTrueKeyInProps = (): keyof VisibilityMap | null => {
    const props = this.props;

    for (const key in VisibilityMap) {
      // @ts-ignore
      if (props[key] === true) {
        return key as keyof VisibilityMap;
      }
    }

    return null;
  };

  private _isVisible = (breakPoint: Breakpoint): boolean => {
    const { only, predicate = reflect } = this.props;

    let isVisible = true;

    if (Array.isArray(only)) {
      isVisible = !only.includes(breakPoint);
    } else if (typeof only === "string") {
      isVisible = only !== breakPoint;
    } else {
      const trueKey = this._getTrueKeyInProps();

      if (trueKey !== null) {
        isVisible = VisibilityMap[trueKey].includes(breakPoint);
      }
    }

    return !predicate(!isVisible);
  };

  private _lastBreakPoint: Breakpoint | null = null;
  private _onResize = () => {
    const width = window.innerWidth;
    const breakPoint = this._getBreakPoint(width);

    if (this._lastBreakPoint === breakPoint) {
      return;
    }

    const lastVisible = this.state.visible;
    const nextVisible = this._isVisible(breakPoint);

    if (lastVisible !== nextVisible) {
      this.setState({ visible: nextVisible });
    }

    this._lastBreakPoint = breakPoint;
  };

  private _onResizeInvoker = () => this._onResizeDebounced();

  render(): any | null {
    if (this.state.visible) {
      const { children } = this.props;

      if (children !== undefined) {
        return children;
      }
    }

    return null;
  }
}
