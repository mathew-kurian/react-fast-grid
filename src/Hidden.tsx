import * as React from "react";
import { Breakpoint, keys } from "./createBreakPoints";
import { breakpoints } from "./gridHelpers";
import debounce from "./debounce";

let keysDsc = [...keys].reverse();
let listeners: (() => void)[] = [];

if (typeof window.matchMedia === "function") {
  for (const [i, key] of keys.entries()) {
    const start = key;
    const end = keys[i + 1];

    const media = window.matchMedia(
      breakpoints.between(start, end).replace("@media", "").trim()
    );

    media.addListener(() => {
      if (process.env.NODE_ENV !== "production") {
        console.log("Fired listener for media items");
      }

      for (const listener of listeners) {
        listener();
      }
    });
  }
} else {
  window.addEventListener(
    "resize",
    () => {
      for (const listener of listeners) {
        listener();
      }
    },
    false
  );
}

interface HiddenState {
  visible: boolean;
}

export interface HiddenProps {
  xsUp?: boolean;
  xsDown?: boolean;
  smUp?: boolean;
  smDown?: boolean;
  mdUp?: boolean;
  mdDown?: boolean;
  lgUp?: boolean;
  lgDown?: boolean;
  xlUp?: boolean;
  xlDown?: boolean;
  only?: Breakpoint | Breakpoint[];
  children?: React.ReactNode;
  debounce?: number;
}

export const isHiddenComponent = (
  element: any
): element is React.ReactElement<HiddenProps> => {
  return (
    element != null &&
    typeof element === "object" &&
    element.type &&
    element.type === Hidden
  );
};

const upFrom = (breakpoint: Breakpoint): Breakpoint[] => {
  return keys.slice(keys.indexOf(breakpoint));
};

const downFrom = (breakpoint: Breakpoint): Breakpoint[] => {
  return keys.slice(0, keys.indexOf(breakpoint) + 1);
};

type UpDownMap = {
  [key in keyof Required<
    Omit<HiddenProps, "children" | "only" | "debounce">
  >]: Breakpoint[];
};

const upDownMap: UpDownMap = {
  xsUp: upFrom("xs"),
  xsDown: downFrom("xs"),
  smUp: upFrom("sm"),
  smDown: downFrom("sm"),
  mdUp: upFrom("md"),
  mdDown: downFrom("md"),
  lgUp: upFrom("lg"),
  lgDown: downFrom("lg"),
  xlUp: upFrom("xl"),
  xlDown: downFrom("xl"),
};

export default class Hidden extends React.Component<HiddenProps, HiddenState> {
  _onResizeDebounced: () => void;

  constructor(props: HiddenProps) {
    super(props);

    const width = window.innerWidth;
    const breakPoint = this._getBreakPoint(width);

    this.state = {
      visible: this._isVisible(breakPoint),
    };

    const { debounce = 5 } = props;

    this._onResizeDebounced = this._getSafeDebounceResize(
      this._onResize,
      debounce
    );
  }

  _getSafeDebounceResize<T extends Function>(
    callback: T,
    timeout: number | undefined
  ): T {
    return typeof timeout === "number" && timeout > 0
      ? debounce(callback, timeout)
      : callback;
  }

  componentWillReceiveProps(nextProps: HiddenProps) {
    const { debounce = 5 } = nextProps;

    if (this.props.debounce !== debounce) {
      this._onResizeDebounced = this._getSafeDebounceResize(
        this._onResize,
        debounce
      );
    }
  }

  componentDidMount() {
    listeners.push(this._onResizeInvoker);
  }

  componentWillUnmount() {
    listeners = listeners.filter(
      (listener) => listener !== this._onResizeInvoker
    );
  }

  _getBreakPoint = (width: number): Breakpoint => {
    const breakPointValues = breakpoints.values;

    for (const key of keysDsc) {
      if (width >= breakPointValues[key]) {
        return key;
      }
    }

    return "xl";
  };

  _getTrueKeyInProp = (): keyof UpDownMap | null => {
    const props = this.props;

    for (const key in upDownMap) {
      // @ts-ignore
      if (props[key] === true) {
        return key as keyof UpDownMap;
      }
    }

    return null;
  };

  _isVisible = (breakPoint: Breakpoint): boolean => {
    const { only } = this.props;

    if (Array.isArray(only)) {
      return only.includes(breakPoint);
    } else if (typeof only === "string") {
      return only === breakPoint;
    }

    const trueKey = this._getTrueKeyInProp();

    if (trueKey !== null) {
      return upDownMap[trueKey].includes(breakPoint);
    }

    return false;
  };

  _lastBreakPoint: Breakpoint | null = null;
  _onResize = () => {
    const width = window.innerWidth;
    const breakPoint = this._getBreakPoint(width);

    if (this._lastBreakPoint === breakPoint) {
      return;
    }

    const lastVisible = this.state.visible;
    const nextVisible = this._isVisible(breakPoint);

    if (lastVisible !== nextVisible) {
      this.setState({ visible: nextVisible });
    }

    this._lastBreakPoint = breakPoint;
  };

  _onResizeInvoker = () => this._onResizeDebounced();

  render() {
    if (this.state.visible) {
      return this.props.children;
    }

    return null;
  }
}
