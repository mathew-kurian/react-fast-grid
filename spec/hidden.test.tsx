import * as React from "react";
import assert from "assert";
import Hidden, { VisibilityMap } from "../lib/Hidden";
import renderer from "react-test-renderer";
import { toMatchDiffSnapshot } from "snapshot-diff";

expect.extend({ toMatchDiffSnapshot });
type Callback = () => void;

class Media {
  static listeners: Callback[] = [];
  addListener(callback: Callback) {
    Media.listeners.push(callback);
  }
}

const setup = (height: number = 1280, width: number = 1024) => {
  //@ts-ignore
  window.innerHeight = height;
  //@ts-ignore
  window.innerWidth = width;
  //@ts-ignore
  window.matchMedia = (mediaString: String): Media => {
    return new Media();
  };
};

const sizes = [
  {
    size: "md",
    width: 1024,
    height: 1280,
  },
  {
    size: "sm",
    width: 768,
    height: 1024,
  },
  {
    size: "xs",
    width: 389,
    height: 512,
  },
  {
    size: "xl",
    width: 2560,
    height: 1440,
  },
  {
    size: "lg",
    width: 1920,
    height: 1080,
  },
];

function helper(a: React.ReactElement, b: React.ReactElement) {
  expect(renderer.create(a).toJSON()).toMatchSnapshot();
  expect(renderer.create(b).toJSON()).toMatchSnapshot();
  expect(renderer.create(a).toJSON()).toMatchDiffSnapshot(
    renderer.create(b).toJSON()
  );
}

for (const { size, width, height } of sizes) {
  assert.deepEqual(VisibilityMap, {
    xsUp: [],
    xsDown: ["sm", "md", "lg", "xl"],
    smUp: ["xs"],
    smDown: ["md", "lg", "xl"],
    mdUp: ["xs", "sm"],
    mdDown: ["lg", "xl"],
    lgUp: ["xs", "sm", "md"],
    lgDown: ["xl"],
    xlUp: ["xs", "sm", "md", "lg"],
    xlDown: [],
  });

  test(`Test viewport.${size} against xs rules`, () => {
    setup(height, width);

    helper(
      <Hidden xsDown>xsDown=true</Hidden>,
      <Hidden only={["xs"]}>xsDown=true</Hidden>
    );

    helper(
      <Hidden xsUp>xsUp=true</Hidden>,
      <Hidden only={["xs", "sm", "md", "lg", "xl"]}>xsUp=true</Hidden>
    );
  });

  test(`Test viewport.${size} against sm rules`, () => {
    setup(height, width);

    helper(
      <Hidden smDown>smDown=true</Hidden>,
      <Hidden only={["xs", "sm"]}>smDown=true</Hidden>
    );

    helper(
      <Hidden smUp>smUp=true</Hidden>,
      <Hidden only={["sm", "md", "lg", "xl"]}>smUp=true</Hidden>
    );
  });

  test(`Test viewport.${size} against md rules`, () => {
    setup(height, width);

    helper(
      <Hidden mdDown>mdDown=true</Hidden>,
      <Hidden only={["xs", "sm", "md"]}>mdDown=true</Hidden>
    );

    helper(
      <Hidden mdUp>mdUp=true</Hidden>,
      <Hidden only={["md", "lg", "xl"]}>mdUp=true</Hidden>
    );
  });

  test(`Test viewport.${size} against lg rules`, () => {
    setup(height, width);

    helper(
      <Hidden lgDown>lgDown=true</Hidden>,
      <Hidden only={["xs", "sm", "md", "lg"]}>lgDown=true</Hidden>
    );

    helper(
      <Hidden lgUp>lgUp=true</Hidden>,
      <Hidden only={["lg", "xl"]}>lgUp=true</Hidden>
    );
  });

  test(`Test viewport.${size} against xl rules`, () => {
    setup(height, width);

    helper(
      <Hidden xlDown>xlDown=true</Hidden>,
      <Hidden only={["xs", "sm", "md", "lg", "xl"]}>lgDown=true</Hidden>
    );

    helper(
      <Hidden xlUp>xlUp=true</Hidden>,
      <Hidden only={["xl"]}>xlUp=true</Hidden>
    );
  });
}
