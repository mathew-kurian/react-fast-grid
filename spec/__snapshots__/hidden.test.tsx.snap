// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test viewport.lg against lg rules 1`] = `"lgDown=true"`;

exports[`Test viewport.lg against lg rules 2`] = `"lgDown=true"`;

exports[`Test viewport.lg against lg rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.lg against lg rules 4`] = `null`;

exports[`Test viewport.lg against lg rules 5`] = `null`;

exports[`Test viewport.lg against lg rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.lg against md rules 1`] = `"mdDown=true"`;

exports[`Test viewport.lg against md rules 2`] = `"mdDown=true"`;

exports[`Test viewport.lg against md rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.lg against md rules 4`] = `null`;

exports[`Test viewport.lg against md rules 5`] = `null`;

exports[`Test viewport.lg against md rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.lg against sm rules 1`] = `"smDown=true"`;

exports[`Test viewport.lg against sm rules 2`] = `"smDown=true"`;

exports[`Test viewport.lg against sm rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.lg against sm rules 4`] = `null`;

exports[`Test viewport.lg against sm rules 5`] = `null`;

exports[`Test viewport.lg against sm rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.lg against xl rules 1`] = `null`;

exports[`Test viewport.lg against xl rules 2`] = `null`;

exports[`Test viewport.lg against xl rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.lg against xl rules 4`] = `null`;

exports[`Test viewport.lg against xl rules 5`] = `null`;

exports[`Test viewport.lg against xl rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.lg against xs rules 1`] = `"xsDown=true"`;

exports[`Test viewport.lg against xs rules 2`] = `"xsDown=true"`;

exports[`Test viewport.lg against xs rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.lg against xs rules 4`] = `null`;

exports[`Test viewport.lg against xs rules 5`] = `null`;

exports[`Test viewport.lg against xs rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.md against lg rules 1`] = `null`;

exports[`Test viewport.md against lg rules 2`] = `null`;

exports[`Test viewport.md against lg rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.md against lg rules 4`] = `"lgUp=true"`;

exports[`Test viewport.md against lg rules 5`] = `"lgUp=true"`;

exports[`Test viewport.md against lg rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.md against md rules 1`] = `null`;

exports[`Test viewport.md against md rules 2`] = `null`;

exports[`Test viewport.md against md rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.md against md rules 4`] = `null`;

exports[`Test viewport.md against md rules 5`] = `null`;

exports[`Test viewport.md against md rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.md against sm rules 1`] = `"smDown=true"`;

exports[`Test viewport.md against sm rules 2`] = `"smDown=true"`;

exports[`Test viewport.md against sm rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.md against sm rules 4`] = `null`;

exports[`Test viewport.md against sm rules 5`] = `null`;

exports[`Test viewport.md against sm rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.md against xl rules 1`] = `null`;

exports[`Test viewport.md against xl rules 2`] = `null`;

exports[`Test viewport.md against xl rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.md against xl rules 4`] = `"xlUp=true"`;

exports[`Test viewport.md against xl rules 5`] = `"xlUp=true"`;

exports[`Test viewport.md against xl rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.md against xs rules 1`] = `"xsDown=true"`;

exports[`Test viewport.md against xs rules 2`] = `"xsDown=true"`;

exports[`Test viewport.md against xs rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.md against xs rules 4`] = `null`;

exports[`Test viewport.md against xs rules 5`] = `null`;

exports[`Test viewport.md against xs rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.sm against lg rules 1`] = `null`;

exports[`Test viewport.sm against lg rules 2`] = `null`;

exports[`Test viewport.sm against lg rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.sm against lg rules 4`] = `"lgUp=true"`;

exports[`Test viewport.sm against lg rules 5`] = `"lgUp=true"`;

exports[`Test viewport.sm against lg rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.sm against md rules 1`] = `null`;

exports[`Test viewport.sm against md rules 2`] = `null`;

exports[`Test viewport.sm against md rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.sm against md rules 4`] = `"mdUp=true"`;

exports[`Test viewport.sm against md rules 5`] = `"mdUp=true"`;

exports[`Test viewport.sm against md rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.sm against sm rules 1`] = `null`;

exports[`Test viewport.sm against sm rules 2`] = `null`;

exports[`Test viewport.sm against sm rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.sm against sm rules 4`] = `null`;

exports[`Test viewport.sm against sm rules 5`] = `null`;

exports[`Test viewport.sm against sm rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.sm against xl rules 1`] = `null`;

exports[`Test viewport.sm against xl rules 2`] = `null`;

exports[`Test viewport.sm against xl rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.sm against xl rules 4`] = `"xlUp=true"`;

exports[`Test viewport.sm against xl rules 5`] = `"xlUp=true"`;

exports[`Test viewport.sm against xl rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.sm against xs rules 1`] = `"xsDown=true"`;

exports[`Test viewport.sm against xs rules 2`] = `"xsDown=true"`;

exports[`Test viewport.sm against xs rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.sm against xs rules 4`] = `null`;

exports[`Test viewport.sm against xs rules 5`] = `null`;

exports[`Test viewport.sm against xs rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xl against lg rules 1`] = `"lgDown=true"`;

exports[`Test viewport.xl against lg rules 2`] = `"lgDown=true"`;

exports[`Test viewport.xl against lg rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xl against lg rules 4`] = `null`;

exports[`Test viewport.xl against lg rules 5`] = `null`;

exports[`Test viewport.xl against lg rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xl against md rules 1`] = `"mdDown=true"`;

exports[`Test viewport.xl against md rules 2`] = `"mdDown=true"`;

exports[`Test viewport.xl against md rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xl against md rules 4`] = `null`;

exports[`Test viewport.xl against md rules 5`] = `null`;

exports[`Test viewport.xl against md rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xl against sm rules 1`] = `"smDown=true"`;

exports[`Test viewport.xl against sm rules 2`] = `"smDown=true"`;

exports[`Test viewport.xl against sm rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xl against sm rules 4`] = `null`;

exports[`Test viewport.xl against sm rules 5`] = `null`;

exports[`Test viewport.xl against sm rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xl against xl rules 1`] = `null`;

exports[`Test viewport.xl against xl rules 2`] = `null`;

exports[`Test viewport.xl against xl rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xl against xl rules 4`] = `null`;

exports[`Test viewport.xl against xl rules 5`] = `null`;

exports[`Test viewport.xl against xl rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xl against xs rules 1`] = `"xsDown=true"`;

exports[`Test viewport.xl against xs rules 2`] = `"xsDown=true"`;

exports[`Test viewport.xl against xs rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xl against xs rules 4`] = `null`;

exports[`Test viewport.xl against xs rules 5`] = `null`;

exports[`Test viewport.xl against xs rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xs against lg rules 1`] = `null`;

exports[`Test viewport.xs against lg rules 2`] = `null`;

exports[`Test viewport.xs against lg rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xs against lg rules 4`] = `"lgUp=true"`;

exports[`Test viewport.xs against lg rules 5`] = `"lgUp=true"`;

exports[`Test viewport.xs against lg rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xs against md rules 1`] = `null`;

exports[`Test viewport.xs against md rules 2`] = `null`;

exports[`Test viewport.xs against md rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xs against md rules 4`] = `"mdUp=true"`;

exports[`Test viewport.xs against md rules 5`] = `"mdUp=true"`;

exports[`Test viewport.xs against md rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xs against sm rules 1`] = `null`;

exports[`Test viewport.xs against sm rules 2`] = `null`;

exports[`Test viewport.xs against sm rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xs against sm rules 4`] = `"smUp=true"`;

exports[`Test viewport.xs against sm rules 5`] = `"smUp=true"`;

exports[`Test viewport.xs against sm rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xs against xl rules 1`] = `null`;

exports[`Test viewport.xs against xl rules 2`] = `null`;

exports[`Test viewport.xs against xl rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xs against xl rules 4`] = `"xlUp=true"`;

exports[`Test viewport.xs against xl rules 5`] = `"xlUp=true"`;

exports[`Test viewport.xs against xl rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xs against xs rules 1`] = `null`;

exports[`Test viewport.xs against xs rules 2`] = `null`;

exports[`Test viewport.xs against xs rules 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Test viewport.xs against xs rules 4`] = `null`;

exports[`Test viewport.xs against xs rules 5`] = `null`;

exports[`Test viewport.xs against xs rules 6`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;
